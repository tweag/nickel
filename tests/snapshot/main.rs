use std::{
    ffi::OsStr,
    path::PathBuf,
    process::{Command, Output},
};
use test_generator::test_resources;

#[test_resources("tests/snapshot/inputs/pretty/*.ncl")]
fn check_pretty_print_snapshots(file: &str) {
    let file = TestFile::from_project_path(file);

    let snapshot = NickelInvocation::new()
        .subcommand("pprint-ast")
        .file(&file)
        .snapshot_stdout();

    insta::assert_snapshot!(file.prefixed_test_name("pretty"), snapshot)
}

#[test_resources("tests/snapshot/inputs/export/*.ncl")]
fn check_export_snapshots(file: &str) {
    let file = TestFile::from_project_path(file);

    let snapshot = NickelInvocation::new()
        .subcommand("export")
        .file(&file)
        .snapshot_stdout();

    insta::assert_snapshot!(file.prefixed_test_name("export"), snapshot)
}

#[test_resources("tests/snapshot/inputs/errors/*.ncl")]
fn check_error_snapshots(file: &str) {
    let file = TestFile::from_project_path(file);

    let snapshot = NickelInvocation::new().file(&file).snapshot_stderr();

    insta::with_settings!({filters => vec![
        // Since error output includes fully-qualified paths to the source file
        // we need to replace those with something static to avoid snapshots
        // differing across machines.
        (r"(?:/.+/tests/snapshot/inputs)", "[INPUTS_PATH]")
    ]},
    {
        insta::assert_snapshot!(file.prefixed_test_name("error"), snapshot)
    });
}

struct TestFile {
    path_buf: PathBuf,
}

impl TestFile {
    fn from_project_path(file: &str) -> Self {
        let mut path_buf = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        path_buf.push(file);
        Self { path_buf }
    }

    fn prefixed_test_name(&self, prefix: &str) -> String {
        let file_name = self
            .path_buf
            .as_path()
            .file_name()
            .and_then(|f| f.to_str())
            .expect("Could not extract file name");
        format!("{}_{}", prefix, file_name)
    }

    fn as_nickel_argument(&self) -> &str {
        self.path_buf
            .to_str()
            .expect("Could not convert path arg to str")
    }
}

struct NickelInvocation {
    cmd: Command,
}

impl NickelInvocation {
    fn new() -> Self {
        let nickel_loc = env!("CARGO_BIN_EXE_nickel");
        let mut cmd = Command::new(nickel_loc);
        cmd.args(["--color", "never"]);
        Self { cmd }
    }

    fn subcommand<S: AsRef<OsStr>>(&mut self, s: S) -> &mut Self {
        self.cmd.arg(s);
        self
    }

    fn file(&mut self, f: &TestFile) -> &mut Self {
        self.cmd.args(["-f", f.as_nickel_argument()]);
        self
    }

    fn run(&mut self) -> Output {
        self.cmd.output().expect("Should be able to capture output")
    }

    fn snapshot_stderr(&mut self) -> String {
        String::from_utf8(self.run().stderr).expect("Output should be utf8")
    }

    fn snapshot_stdout(&mut self) -> String {
        String::from_utf8(self.run().stdout).expect("Output should be utf8")
    }
}
