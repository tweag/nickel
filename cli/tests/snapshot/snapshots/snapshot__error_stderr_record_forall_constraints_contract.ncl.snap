---
source: tests/snapshot/main.rs
expression: err
---
error: contract broken by the caller: field not allowed in tail: `x`
  ┌─ [INPUTS_PATH]/errors/record_forall_constraints_contract.ncl:3:19
  │
3 │ let f | forall r. { ; r } -> { x: Number; r } = fun r => %record_insert% "x" r 1 in f { x = 0 }
  │                   -------                                                             --------- evaluated to this expression
  │                   │                                                                    
  │                   expected type of the argument provided by the caller
  │
  ┌─ <unknown> (generated by evaluation):1:1
  │
1 │ { ... }
  │ ------- evaluated to this value
  │
  = This error may happen in the following situation:
        1. A function `f` is bound by a contract: e.g. `Number -> Number`.
        2. `f` is called with an argument of the wrong type: e.g. `f false`.
  = Either change the contract accordingly, or call `f` with an argument of the right type.
  = Note: this is an illustrative example. The actual error may involve deeper nested functions calls.

note: 
  ┌─ [INPUTS_PATH]/errors/record_forall_constraints_contract.ncl:3:58
  │
3 │ let f | forall r. { ; r } -> { x: Number; r } = fun r => %record_insert% "x" r 1 in f { x = 0 }
  │                                                          ^^^^^^^^^^^^^^^^^^^^^^^ While calling to r

note: 
  ┌─ [INPUTS_PATH]/errors/record_forall_constraints_contract.ncl:3:85
  │
3 │ let f | forall r. { ; r } -> { x: Number; r } = fun r => %record_insert% "x" r 1 in f { x = 0 }
  │                                                                                     ----------- (1) calling f

note: 
  ┌─ [INPUTS_PATH]/errors/record_forall_constraints_contract.ncl:3:49
  │
3 │ let f | forall r. { ; r } -> { x: Number; r } = fun r => %record_insert% "x" r 1 in f { x = 0 }
  │                                                 -------------------------------- (2) calling <func>


